File: Leetcode/C++/1038. Binary Search Tree to Greater Sum Tree/solution.cpp
// https://leetcode.com/problems/binary-search-tree-to-greater-sum-tree/

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution // DFS
{
public:
    int total = 0;
    TreeNode *bstToGst(TreeNode *root)
    {
        if (root == NULL)
        {
            return new TreeNode(0);
        }

        TreeNode *right = bstToGst(root->right);
        total += right->val;
        root->val = total;
        bstToGst(root->left);
        return root;
    }
};

File: Leetcode/Typescript/1038. Binary Search Tree to Greater Sum Tree/solution.ts
// https://leetcode.com/problems/binary-search-tree-to-greater-sum-tree/description/

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function bstToGst(root: TreeNode | null): TreeNode | null {
  let total = 0;

  function dfs(root) {
    if (!root) {
      return;
    }
    dfs(root.right);
    total += root.val;
    root.val = total;
    dfs(root.left);
  }

  dfs(root);
  return root;
}

// Recursion
function bstToGst(root: TreeNode | null): TreeNode | null {
  function dfs(root: TreeNode | null, total: number): number {
    if (!root) {
      return total;
    }

    const right = dfs(root.right, total);
    root.val += right;

    return dfs(root.left, root.val);
  }

  dfs(root, 0);
  return root;
}
